DigitalLearningSolutions.Web.Tests/Services/EmailVerificationServiceTests.cs

```diff
@@ -0,0 +1,168 @@
+    using DigitalLearningSolutions.Data.Utilities;
+    using DigitalLearningSolutions.Web.Services;
+    using FakeItEasy;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class EmailVerificationServiceTests
+    {
+        private IEmailVerificationDataService emailVerificationDataService = null!;
+        private IEmailService emailService = null!;
+        private IClockUtility clockUtility = null!;
+        private EmailVerificationService emailVerificationService = null!;
+
+        [SetUp]
+        public void Setup()
+        {
+            emailVerificationDataService = A.Fake<IEmailVerificationDataService>();
+            emailService = A.Fake<IEmailService>();
+            clockUtility = A.Fake<IClockUtility>();
+            emailVerificationService = new EmailVerificationService(
+                emailVerificationDataService,
+                emailService,
+                clockUtility
+            );
+        }
+
+        [Test]
+        [TestCase(false)]
+        [TestCase(true)]
+        public void AccountEmailIsVerifiedForUser_returns_expected_result(bool expectedResult)
+        {
+            // Given
+            const int userId = 2;
+            const string email = "test@email.com";
+            A.CallTo(() => emailVerificationDataService.AccountEmailIsVerifiedForUser(userId, email))
+                .Returns(expectedResult);
+
+            // When
+            var result = emailVerificationService.AccountEmailIsVerifiedForUser(userId, email);
+
+            // Then
+            result.Should().Be(expectedResult);
+        }
+
+        [Test]
+        public void
+            CreateEmailVerificationHashesAndSendVerificationEmails_does_not_send_emails_if_no_email_requires_verification()
+        {
+            // Given
+            var userAccount = UserTestHelper.GetDefaultUserAccount();
+
+            // When
+            emailVerificationService.CreateEmailVerificationHashesAndSendVerificationEmails(
+                userAccount,
+                new List<PossibleEmailUpdate>(),
+                "example.com"
+            );
+
+            // Then
+            A.CallTo(
+                () => emailVerificationDataService.UpdateEmailVerificationHashIdForPrimaryEmail(
+                    A<int>._,
+                    A<string>._,
+                    A<int>._
+                )
+            ).MustNotHaveHappened();
+
+            A.CallTo(
+                () => emailVerificationDataService.UpdateEmailVerificationHashIdForCentreEmails(
+                    A<int>._,
+                    A<string?>._,
+                    A<int>._
+                )
+            ).MustNotHaveHappened();
+
+            A.CallTo(() => emailService.SendEmail(A<Email>._)).MustNotHaveHappened();
+        }
+
+        [Test]
+        public void
+            CreateEmailVerificationHashesAndSendVerificationEmails_updates_hashId_and_sends_email_when_primary_email_requires_verification()

```

Comment:
`CreateEmailVerificationHashesAndSendVerificationEmails` doesn't actually care whether you're updating a primary or centre email, so I don't think you should be referencing them in the test names, since it is not possible to distinguish between what emails are updating in the Given/When sections. What you should be testing is:
* `SendEmail` and both `UpdateEmailVerificationHashId` methods are called if `unverifiedEmails` is not empty
* Each `UpdateEmailVerificationHashId` is only called once for each distinct email address in `unverifiedEmails`
* `SendEmail` is only called once for each distinct email address in `unverifiedEmails`